src/qforte/utils/qbt_tapering.py

```diff
@@ -0,0 +1,321 @@
+                prt_chck_mtrx[i, n_qubits - trgt - 1] = 1
+            if XYZ_gate in ['X', 'Y']:
+                prt_chck_mtrx[i, 2 * n_qubits - trgt - 1] = 1
+
+    # 2) Find a basis of the kernel of the parity check matrix
+    # See https://en.wikipedia.org/wiki/Kernel_(linear_algebra)#Computation_by_Gaussian_elimination
+
+    ## Augment the (n_strings) × (2 * n_qubits) parity check matrix by the (2 * n_qubits) × (2 * n_qubits) identity matrix
+
+    prt_chck_mtrx_aug = np.concatenate((prt_chck_mtrx, np.identity(2 * n_qubits, dtype=np.uint8)), axis=0)
+
+    ## Find column echelon form (CEF) of augmented matix.
+
+    ### To speed up the process, we only work with the part of the parity check matrix that has not been transformed yet to CEF.
+    ### This is controlled by the "idx" variable
+    idx = 0
+
+    for row in range(n_strings):
+        ### Find a non-zero row
+        clmns_not_in_CEF = prt_chck_mtrx_aug[row,idx:]
+        if np.any(clmns_not_in_CEF):
+            ### Find positions of non-zero elements
+            nonzero_indices = np.argwhere(clmns_not_in_CEF)
+            column = int(nonzero_indices[0]) + idx
+            ### Swap columns to bring first non-zero element in this row immediately to the right of that of the previous row
+            prt_chck_mtrx_aug[:, [column , idx]] = prt_chck_mtrx_aug[:, [idx, column]]
+            ### Eliminate remaining non-zero elements in the submatrix
+            for i in range(1,nonzero_indices.shape[0]):
+                column2 = int(nonzero_indices[i]) + idx
+                prt_chck_mtrx_aug[:,column2] = np.bitwise_xor(prt_chck_mtrx_aug[:, idx], prt_chck_mtrx_aug[:,column2])
+            ### decrease the number of columns that need to be checked
+            idx += 1
+
+    ## Find basis vectors of kernel by checking for zero columns in the CEF of the parity check matrix
+
+    dim = 0
+    basis = np.zeros((1, 2 * n_qubits), dtype=np.uint8)
+    for column in range(2 * n_qubits):
+        row = n_strings
+        if not np.any(prt_chck_mtrx_aug[:row, column]):
+            if dim == 0:
+                basis = prt_chck_mtrx_aug[row:, column].reshape((1,2 * n_qubits))
+                dim += 1
+            else:
+                basis = np.concatenate((basis, prt_chck_mtrx_aug[row:, column].reshape((1,2 * n_qubits))), axis=0)
+
+    # 3) Find maximal Abelian subgroup of the basis of ker{E}
+    # This is done using the symplectic bilinear form B that the binary vector space is equipped with:
+    # B(v,w) = v * B * w.T,
+    # where v and w are binary vectors and B is the matrix of the bilinear form.
+
+    ## Construct matrix of bilinear form.
+
+    top = np.concatenate((np.zeros((n_qubits, n_qubits), dtype=np.uint8), np.identity(n_qubits, dtype=np.uint8)), axis=1)
+    bottom = np.concatenate((np.identity(n_qubits, dtype=np.uint8), np.zeros((n_qubits, n_qubits), dtype=np.uint8)), axis=1)
+    blnr_frm_mtrx = np.concatenate((top, bottom), axis=0)
```

Comment:
Numpy's [block](https://numpy.org/doc/stable/reference/generated/numpy.block.html) should make what you're trying to do more transparent.